terratest/test/k8gb_abstract_full_roundrobin_test.go

```diff
@@ -0,0 +1,120 @@
+)
+
+func AbstractTestFullRoundRobin(t *testing.T, n int) {
+	if n < 2 || n > 8 {
+		t.Logf("Use value of n that represents the number of clusters from interval [2,8]")
+		t.FailNow()
+	} else {
+		t.Logf(fmt.Sprintf("Running TestFullRoundRobin for %d clusters", n))
+	}
+	tags := []string{"eu", "us", "cz", "af", "ru", "ap", "uk", "ca"}
+	var instances []*utils.Instance
+
+	t.Parallel()
+	const host = "roundrobin-test.cloud.example.com"
+	const gslbPath = "../examples/roundrobin2.yaml"
+
+	// start all the test apps on all the clusters
+	for i := 0; i < n; i+=1 {
+		instance, er := utils.NewWorkflow(t, fmt.Sprintf("k3d-test-gslb%d", i+1), 5053 + i).
+			WithGslb(gslbPath, host).
+			WithTestApp(tags[i]).
+			Start()
+		require.NoError(t, er)
+		instances = append(instances, instance)
+		defer instance.Kill()
+	}
+	var err error
+	t.Run(fmt.Sprintf("round-robin on %d concurrent clusters with podinfo running", n), func(t *testing.T) {
+		for _, ins  := range instances {
+			err = ins.WaitForAppIsRunning()
+			require.NoError(t, err)
+		}
+	})
+
+	// at the beginning, it should contain all the targets
+	var workingTargets []string
+	for _, ins  := range instances {
+		workingTargets = append(workingTargets, ins.GetLocalTargets()...)
+	}
+	t.Run(fmt.Sprintf("all %d clusters should be interconnected", n), func(t *testing.T) {
+		allShouldExpectTheseTargets(t, instances, workingTargets)
+	})
+
+	// kill the apps on clusters one by one and expect less and less targets to be available
+	for i, instance := range instances {
+		t.Run(fmt.Sprintf("kill podinfo on cluster %d (%s)", i + 1, tags[i]), func(t *testing.T) {
+			workingTargets = dedupe(subtract(workingTargets, instance.GetLocalTargets()))
+			t.Logf("New expected targets: %v", workingTargets)
+			instance.StopTestApp()
+			allShouldExpectTheseTargets(t, instances, workingTargets)
+		})
+	}
+
+	// start the test apps again on each cluster and check if the targets start appearing
+	for i, instance := range instances {
+		t.Run(fmt.Sprintf("start podinfo on cluster %d (%s)", i + 1, tags[i]), func(t *testing.T) {
+			instance.StartTestApp()
+			workingTargets = dedupe(append(workingTargets, instance.GetLocalTargets()...))
+			t.Logf("New expected targets: %v", workingTargets)
+			allShouldExpectTheseTargets(t, instances, workingTargets)
+		})
+	}
+}
+
+func allShouldExpectTheseTargets(t *testing.T, instances []*utils.Instance, workingTargets []string) {
+	for _, instance := range instances {
+		err := instance.WaitForExpected(workingTargets)
+		require.NoError(t, err)
+	}
+}
+
+// helper function for subtracting slice from slice (in O(n))
+func subtract(from, what []string) (diff []string) {
+	aux := make(map[string]bool, len(what))
+	for _, w := range what {
+		aux[w] = true
+	}
+	for _, f := range from {
+		if _, found := aux[f]; !found {
+			diff = append(diff, f)
+		}
+	}
+	return
+}
+
+// helper function for de-duplicating elements in slice (in O(n))
+func dedupe(s []string) (uniq []string) {
```

Comment:
please rename the function to something more meaningful  (e.g. in SQL wa have distinct keyword for that)

Reply:
changed to `distinct`